onedal/decomposition/pca.py

```diff
@@ -15,30 +15,77 @@
 
 import numpy as np
+from sklearn.decomposition._pca import _infer_dimension
+from sklearn.utils.extmath import stable_cumsum
 
 from daal4py.sklearn._utils import sklearn_check_version
 
 from ..common._base import BaseEstimator
-from ..common.hyperparameters import get_hyperparameters
 from ..datatypes import _convert_to_supported, from_table, to_table
 
 
 class PCA(BaseEstimator):
     def __init__(
-        self, n_components=None, is_deterministic=True, method="precomputed", copy=True
+        self,
+        n_components=None,
+        is_deterministic=True,
+        method="cov",
+        whiten=False,
     ):
         self.n_components = n_components
         self.method = method
         self.is_deterministic = is_deterministic
+        self.whiten = whiten
 
-    def get_onedal_params(self, data):
+    def get_onedal_params(self, data, stage=None):
+        if stage is None:
+            n_components = self._resolve_n_components_for_training(data.shape)
+        elif stage == "predict":
+            n_components = self.n_components_
         return {
             "fptype": "float" if data.dtype == np.float32 else "double",
             "method": self.method,
-            "n_components": self.n_components,
+            "n_components": n_components,
             "is_deterministic": self.is_deterministic,
+            "whiten": self.whiten,
         }
 
+    def _resolve_n_components_for_training(self, shape_tuple):
+        if self.n_components is None or self.n_components == "mle":
+            return min(shape_tuple)
+        elif (
+            isinstance(self.n_components, float)
+            and self.n_components > 0.0
+            and self.n_components <= 1.0
+        ):
+            return min(shape_tuple)
+        else:
+            return self.n_components
+
+    def _resolve_n_components_for_result(self, shape_tuple):
+        if self.n_components is None:
+            return min(shape_tuple)
+        elif self.n_components == "mle":
+            return _infer_dimension(self.explained_variance_, shape_tuple[0])
+        elif 0.0 < self.n_components < 1.0:
+            ratio_cumsum = stable_cumsum(self.explained_variance_ratio_)
+            return np.searchsorted(ratio_cumsum, self.n_components, side="right") + 1
+        elif isinstance(self.n_components, float) and self.n_components == 1.0:
+            return min(shape_tuple)
+        else:
+            return self.n_components
+
+    def _compute_noise_variance(self, n_components, n_sf_min):
+        if n_components < n_sf_min:
+            if len(self.explained_variance_) == n_sf_min:
+                return self.explained_variance_[n_components:].mean()
+            elif len(self.explained_variance_) < n_sf_min:
+                resid_var = self.variances_.sum()
+                resid_var -= self.explained_variance_.sum()
+                return resid_var / (n_sf_min - n_components)
+        else:
```

Comment:
what should be returned if len(self.explained_variance_) > n_sf_min? or is it not feasible? If its meant to be zero then should drop the else and just return 0.0 after the end of if statement

Reply:
yes, the case is not feasible